harvester/harvest/alma.py

```diff
@@ -78,51 +89,94 @@ def parse_marc_records_from_files(self) -> Iterator[MARCRecord]:
             - run_type=daily
         """
-        return HARVEST_TYPE_MAP[self.harvest_type] in filepath
+        return [
+            filepath
+            for filepath in filepaths
+            if HARVEST_TYPE_MAP[self.harvest_type] in filepath
+        ]
+
+    def _filter_filepaths_by_dates(self, filepaths: list[str]) -> list[str]:
+        """Filter list of XML files by date.
 
-    def _filter_filepath_by_dates(self, filepath: str) -> bool:
-        """Bool if from and/or until dates align with filepath dates.
+        Given a list of XML files, the method will retrieve the date (YYYY-MM-DD)
+        from the filepath and determine whether the date is valid by checking
+        if the date falls within the specified 'from_date' and/or 'until_date'
+        arguments passed to the harvester.
 
         Example filepath: alma-2024-03-01-daily-extracted-records-to-index_19.xml
             - run_date=2024-03-01
         """
-        match = FILEPATH_DATE_REGEX.match(filepath)
-        if not match:  # pragma: nocover
-            message = f"Could not parse date from filepath: {filepath}"
-            logger.warning(message)
-            return False
+        filtered_filepaths = []
+        for filepath in filepaths:
+            filepath_date = convert_to_utc(
+                date_parser(self._get_date_from_filepath(filepath))  # type: ignore[arg-type]
+            )
+            if self._validate_filepath_date(filepath_date):
+                filtered_filepaths.append(filepath)
+        return filtered_filepaths
+
+    def _validate_filepath_date(self, date: datetime.datetime) -> bool:
+        """Check if a given date is valid.
 
-        filepath_date = convert_to_utc(date_parser(match.group(1)))
-        if self.from_datetime_object and filepath_date < self.from_datetime_object:
-            return False
-        if self.until_datetime_object and filepath_date >= self.until_datetime_object:
-            return False
+        Dates are valid if it meets at least one of the following condition:
 
-        return True
+        * Date is (a) on or after MITAlmaHarvester.from_date AND (b) before
+            MITAlmaHarvester.until_date if 'from_date' and 'until_date' is specified.
+        * Date is on or after MITAlmaHarvester.from_date if only 'from_date' is specified.
+        * Date is before MITAlmaHarvester.until_date if only 'until_date' is specified.
+        """
+        if (self.from_datetime_object and self.until_datetime_object) and (
+            date >= self.from_datetime_object and date < self.until_datetime_object
+        ):
+            return True
+        if (self.from_datetime_object and self.until_datetime_object is None) and (
+            date >= self.from_datetime_object
+        ):
+            return True
+        if (self.until_datetime_object and self.from_datetime_object is None) and (
+            date < self.until_datetime_object
+        ):
+            return True
+        return False
```

Comment:
I might propose combining methods `_filter_filepaths_by_dates()` and `_validate_filepath_date()`, for a couple of reasons.

1- I'm not sure `_validate_filepath_date()` feels right.  It feels less about validating, and more about **filtering** when a particular condition is true (i.e. the date in the filepath meets `self.from_date` and/or `self.until_date` criteria).

2- If you consider a rename... then it starts to sound like the other, "parent" method `_filter_filepaths_by_dates()`!

3- I think the logic in `_validate_filepath_date()` could be simplifed from multiple `if` statements to a single one.  And with that done, then it feels like it could drop in to the parent method more easily.

Something along the lines of:

```python
def _filter_filepaths_by_dates(self, filepaths: list[str]) -> list[str]:
    """Filter list of XML files by date.

    Given a list of XML filepaths, the method will retrieve the date (YYYY-MM-DD)
    from the filename and determine whether the file should be included by checking
    if the extracted date falls within the specified 'from_date' and/or 'until_date'
    arguments passed to the harvester.

    Example filepath: alma-2024-03-01-daily-extracted-records-to-index_19.xml
        - run_date=2024-03-01
    """
    filtered_filepaths = []
    for filepath in filepaths:
        if filepath_date_string := self._get_date_from_filepath(filepath):
            filepath_date = date_parser(filepath_date_string)

            # include where filepath date meets harvester from/until date criteria
            if (
                self.from_datetime_object is None
                or filepath_date >= self.from_datetime_object
            ) and (
                self.until_datetime_object is None
                or filepath_date < self.until_datetime_object
            ):
                filtered_filepaths.append(filepath)

    return filtered_filepaths
```

Note that this would also require `_get_date_from_filepath()` returning `None` if it cannot extract a date, to support this conditional above.

Reply:
Mentioned in the PR notes @jonavellecuerdo, but please pushback if you're feeling like it should remain two methods.  If so, maybe we'd benefit form renaming `_validate_filepath_date()` to something like `_is_date_within_harvest_date_scope()` or something, given that it's _only_ getting  date; it really doesn't need to know it's got anything to do with filenames.

Reply:
@ghukill Just to confirm--the proposed code above would include `filepath` if: `self.from_datetime_object is None and self.until_datetime_object is None`, right? 

I think this makes sense for incremental harvests. However, for full harvests, [`_get_latest_run_date()` will raise an error if the latest full run date could not be determined](https://github.com/MITLibraries/geo-harvester/pull/271/files#diff-2e6b91e7e7f294d31aaf65ae9da8b0fdb4e602cfd1cad4e951ddcb8aa68e0bbbR104-R105)--i.e., for Alma full harvests, `from_date` is technically required (because otherwise, it would get all "full-extracted" and all "daily-extracted" filepaths. Do you see any potential issues with this? 🤔

Reply:
@jonavellecuerdo - added this comment, https://github.com/MITLibraries/geo-harvester/pull/271/commits/898e3b22ccb542df3eb274579267a1068cbddebd#r1563099498, which may touch on this?

Reply:
Updated to consolidate methods into one and rebased commits.